### ISSUE TITLE
Generalize tests that use the agent abstraction

### ISSUE THREAD

**Issue by dpino** [tc39/test262#1351]
I'm currently working on building up an environment that could allow to run test262 within a browser (as web-platform-test). The idea is to do this process as most automatic as possible without rewriting test262's tests as much as possible.

So far I've been exploring the tests that use the agent abstraction. In a browser context the agent abstraction can be a DedicatedWorker, SharedWorker, ServiceWorker, Window or IFrame's contentWindow. TC39's agent abstraction has a different API that workers. Although it would be possible to write harnessing code that serves as a bridge between agent an a specific worker, it's likely that these tests using agent will need to get some parts rewritten.

Specifically, isolating workers code, parametrizing timeouts and getting rid or rework the sleep calls are the most important issues. Here is some code with my attempt so far to make these tests to work as WPT tests https://github.com/dpino/gecko-dev/pull/19b

**Comment by lars-t-hansen** [comment]
Yeah, the agent abstraction is pretty clunky, and I've also tried to move it to browsers without much luck.  The reason for the clunkiness is in large part that different JS shells currently have such different ideas about how to create and manage and communicate among workers.

I don't know how much hurry you're in to do this.  Over in https://github.com/WebAssembly/threads/issues/52 there's an effort going on to move the HTML Worker abstraction into all the JS shells (at least in part), so that it will be simpler to share test cases and to avoid the agent abstraction mess.  The target there is WebAssembly but JS should benefit directly from the work.  I don't have any kind of ETA on that work; I think we all more or less agree on the API, now we just have to go and implement it.

**Comment by annevk** [comment]
@lars-t-hansen our idea was to wrap the tests so they can be run using the web-platform-tests framework. So instead of a fake window or worker environment, we get to run the tests with actual window/worker/shared worker/service worker/audio worklet semantics. Just window or worker is not sufficient I think.

**Comment by lars-t-hansen** [comment]
@annevk, sure, you can do whatever you want :)  All I'm saying is that if you can afford to wait for the JS shells to have a more or less standard Worker API then we can probably rewrite the test262 agent abstraction so that it is less hostile to the browsers.  But otherwise I'm fine with anything.

**Comment by dpino** [comment]
I think this approach would work. At first I was concerned about not being able to run new workers inside the test plus the calls that explicitly enter and exit the UI event loop (a window in browsers).

With regard to the first issue, it's not a problem since the current tests that use the agent API do not instantiate new workers (it's just us wanting to run these tests from different types of workers). Then with regard to the `enterEventLoop` and `exitEventLoop` calls, although the browser's Worker API do not feature these functions, these calls could be implemented as empty when running the tests as WPT.

I think what I mean gets better illustrated with an example:

**wait/nan-timeout.js**

```javascript
// Copyright (C) 2017 Mozilla Corporation.  All rights reserved.
// This code is governed by the BSD license found in the LICENSE file.

/*---
esid: sec-atomics.wait
description: >
  Test that Atomics.wait does not time out with a NaN timeout
---*/

$262.agent.start(
`
$262.agent.receiveBroadcast(function (sab, id) {
  var ia = new Int32Array(sab);
  $262.agent.report(Atomics.wait(ia, 0, 0, NaN));  // NaN => Infinity
  $262.agent.leaving();
})
`);

var ia = new Int32Array(new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT));

$262.agent.broadcast(ia.buffer);
$262.agent.sleep(500);                // Ample time
assert.sameValue($262.agent.getReport(), null);
Atomics.wake(ia, 0);
assert.sameValue(getReport(), "ok");

function getReport() {
    var r;
    while ((r = $262.agent.getReport()) == null)
        $262.agent.sleep(100);
    return r;
}
```

This test262 test would rewritten as this using the new JS Shell Worker API.

```javascript
var w = new Worker(`
  onmessage = function(e) {
    var sab = e.data[0];
    var ia = new Int32Array(sab);
    postMessage(Atomics.wait(ia, 0, 0, NaN));  // NaN => Infinity
  }
`);
w.onmessage = function(e) {
    assert.sameValue(e.data, "ok");
    exitEventLoop();
}

var ia = new Int32Array(new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT));
w.postMessage([ia.buffer]);
Atomics.wake(ia, 0);
enterEventLoop();
```

This test can be run from a browser (either using a Window or a different type of Worker) almost directly. Some concerns though.

AFAIK the Workers API can only create workers out of a file (`new Worker(filename)`). I believe it's more convenient to embed a worker code in the same test file as it is now though. So perhaps instead of a `new Worker(source)` call, having a `Worker.create(source)` constructor will be more convenient. It will be relatively easy to provide this constructor in a browser context using a polyfill (that would use a BlobBuilder under the hood to create the worker from a string).

Then with regard to the test above, I removed the `sleep` call but I'm not sure whether it should still be there. Generally all these agent tests feature a sleep call somewhere to give a chance to the worker to run and wait, then in the caller thread a wake is called after the sleeping time. What will happen with the sleep calls using the new API?

**Comment by leobalter** [comment]
I'm fine to add/rewrite/refactor any abstraction that helps the communication between the tests and implementations. My only concern is favoring some specific implementation, so we can keep Test262 host (and agent) agnostic and some compatibility can always be improved.

We've had some of this value applied to detaching ArrayBuffers, where the current harness just throws an error mentioning the required implementation. My first try for that was a fall back to the non-standard ArrayBuffer.transfer if available.

In any case, I believe the suggested harness can provide another abstraction level where you don't have to use close to real use cases as  `new Worker` but a top layer that defers to that.

The idea of tests deferring to abstractions - agent or harness files - means these can be replaced by anything consuming it, including the web platform tests.

**Comment by dpino** [comment]
I forked Lar's https://github.com/dpino/Worker repository and adapt the current tests that use agent with the proposed Worker implementation for JavaScript shells. The changes are available here: https://github.com/dpino/Worker/tree/port-262-agent-tests. To run the tests, I run them directly in SpiderMonkey (built with path that adds putMessage/getMessage applied).

Then I took one of the tests (nan-timeout.js) and run it in a browser to check whether any changes are required. I found that minimal changes were required. Besides the difference between `new Worker` in the JS Shell and the browser, I found out that I needed to terminate the workers on `exitEventLoop` (if I skip w.terminate in the shell, the script gets stalled; OTOH, if I left it as it's the test doesn't run OK within a browser).

Overall I like this new abstraction more than the previous one.

**Comment by lars-t-hansen** [comment]
Yes, it's unfortunate that `new Worker` in the shell takes source code while `new Worker` in the browser takes a URL (and I think most browsers are removing the possibility for that to be a data: URL, or we could hack something together fairly easily).

I expect all of the shells have a read-text-from-file function so that it will be possible to abstract this with a little preamble, and always have `new Worker` in the test cases take a relative URL, ie, a file name in the shells.  The preamble will be shell-specific however.  Maybe we should look into that too.

```
let _Worker = Worker;
Worker = function (fn) {
  return new _Worker(os.file.read(fn));
}
```

Re the termination semantics, I agree this is also unfortunate.  I think some experimentation is called for, since the browser does not need workers to be terminated explicitly.  Maybe a top-level function `terminateWorkers()` that takes an array of workers and terminates all of them; in the browser it would do nothing.  (In the SpiderMonkey shell, calling `quit(0)` terminates the process, workers and all, so that type of exit could also be used.  Of course, it's not called `quit` everywhere.)

**Comment by annevk** [comment]
`data:` URLs will continue to be allowed, though they will make the worker run in an opaque origin. I don't think that would have any drawbacks on JavaScript though and `postMessage()` and `SharedArrayBuffer` should continue to work. However, `data:` URLs don't work for service workers so are not really a solution.

If you could clearly identify the worker code somehow we could just pre-process it with Python, as part of the web-platform-tests wrapper.

**Comment by lars-t-hansen** [comment]
I wonder, if the worker based on a data: URL runs in an opaque origin then can that worker receive shared memory from the main thread?  Memory can only be shared within an agent cluster; IIUC the agent cluster mapping for the web essentially restricts agents within the cluster to be same-origin (modulo document.origin manipulation).

**Comment by annevk** [comment]
I don't think as currently specified "sandboxed globals" end up in a different cluster. E.g., `<iframe srcdoc="<script>...</script>" sandbox>` can share memory, but I'm not sure if we've tested that.

**Comment by dpino** [comment]
My thoughts on this issue.

I've rewritten the current tests that use the agent abstraction using Lars proposed new API. Tests are available here [1]. I already commented on that some time ago actually, the difference is that now I also made all those tests to run within Firefox and I also imported them as web-platform-tests [2]. It's necessary a few additional changes (besides those ones I commented, some calls should be a setTimeout rather than sleep to make the tests run successfully in a browser). By importing the tests as WPT, I covered the case where the caller is an IFrame, Window, DedicatedWorker and SharedWorker and the callee is a DedicatedWorker.

Now regarding parameterization of agents in test262. It's my understanding that what we are discussing is to run one single test with all possible agent combinations, is that correct? If we want to do that the con is that test262 tests only expect one kind of result (specified in the YAML header of the test).

For instance, consider the `wait/nan-timeout.js` test:

```
let w = new Worker(`
    onmessage = function(e) {
        let ia = e.data[0];
        let ret = Atomics.wait(ia, 0, 0, NaN); // NaN => Infinity
        postMessage([ret]);
    }
`);

w.onmessage = function(e) {
    assert.sameValue(e.data[0], "ok");
    exitEventLoop();
}

let ia = new Int32Array(new SharedArrayBuffer(Int32Array.BYTES_PER_ELEMENT));
w.postMessage([ia]);

// Ample time
sleep(500).then(() => Atomics.wake(ia, 0));

enterEventLoop();
w.terminate();
```
If the Worker takes the role of a DedicatedWorker, the test passes as expected. However, if the Worker takes the role of a SharedWorker the test throws an Exception (TypeError: SharedArrayBuffer cannot be cloned in this context). That's the expected behavior according to the HTML spec, thus the test should add a 'negative' clause on its header. Something like:

/*---
esid: sec-atomics.wait
description: >
  Test that Atomics.wait does not time out with a NaN timeout
negative:
   phase: runtime
   type: TypeError
---*/

Likewise if the Worker is an IFrame the test will throw an Error (TypeError: waiting is not allowed on this thread). In this case the error message is different although the type of error is the same.

To me it makes sense to extend the agent API to allow different behavior parametrization (the agent behaves like a SharedWorker, like an IFrame, etc) but I don't see why it would we useful to run a test that was coded with a type of worker in mind (in the case of `nan-timeout.js` the test validates whether is possible to pass a NaN value to `Atomics.wait` and it's coded as if it were a DedicatedWorker) and run it in all types of Workers. If test262 needs a test that checks whether it's not possible to pass a SAB to a SharedWorker I think a separate test for that plus supporting Agent parameterization would do the work.

Opinions? thoughts? @annevk @lars-t-hansen

**Comment by annevk** [comment]
I think it's fine, maybe even preferred, that if a test is coded for agents with their fields set to particular values, we only run the test for all combinations of matching HTML agents. So if a test requires a particular agent's [[CanBlock]] to be true, we don't use a similar-origin window agent for it. Does that make sense?

**Comment by annevk** [comment]
Ah yeah, does test262 even consider cross-cluster tests? (I also hope the metadata is sufficiently flexible so that when the [[CanBlock]] field doesn't matter it's allowed to be either.)

**Comment by dpino** [comment]
SGTM. I think test262 doesn't handle cross-cluster tests, but I brought that up thinking of the [cannot-block] and shared-worker scenario.

So if only same cluster case is needed to handle when running as the tests as WPT, that means the agent will always be a DedicatedWorker and in some cases an IFrame's window or a child window. All good then, I need to extend the tests to cover the iframe and child window cases.
